src/backend/brokers/go.emails/encrypt.go

```diff
@@ -0,0 +1,148 @@
+package email_broker
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/base64"
+	. "github.com/CaliOpen/Caliopen/src/backend/defs/go-objects"
+	log "github.com/Sirupsen/logrus"
+	"github.com/emersion/go-message"
+	"math/rand"
+	"net/mail"
+	"strings"
+	"time"
+)
+
+var boundaryChars = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789+:=?")
+
+func RandomString(n int) string {
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = boundaryChars[rand.Intn(len(boundaryChars))]
+	}
+	return string(b)
+}
+
+// Return a MIME boudary as specified in RFC 1341 (7.2.1)
+//
+// boundary := 0*69<bchars> bcharsnospace
+//
+// bchars := bcharsnospace / " "
+//
+// bcharsnospace :=    DIGIT / ALPHA / "'" / "(" / ")" / "+"  /
+// "_"
+//                / "," / "-" / "." / "/" / ":" / "=" / "?"
+func (b *EmailBroker) NewBoundary() string {
+	return RandomString(42)
+}
+
+func (b *EmailBroker) NewMessageId(uuid []byte) string {
+	// sha256 internal message id to form external message id
+	hasher := sha256.New()
+	hasher.Write(uuid)
+	sha := base64.URLEncoding.EncodeToString(hasher.Sum(nil))
+	messageId := sha + "@" + b.Config.PrimaryMailHost // should be the default domain in case there are multiple 'from' addresses
+	return messageId
+}
+
+//Marshal an encrypted email
+func (b *EmailBroker) MarshalEncryptedEmail(msg *Message, em *EmailMessage) (err error) {
+
+	mainHeader := make(message.Header)
+	params := map[string]string{"boundary": b.NewBoundary(), "protocol": "application/pgp-encrypted"}
+	mainHeader.Set("Subject", msg.Subject)
+	mainHeader.Set("Date", time.Now().Format(time.RFC1123Z))
+	mainHeader.SetContentType("multipart/encrypted", params)
+	mainHeader.Set("X-Mailer", "Caliopen-"+b.Config.AppVersion)
+
+	addr_fields := formatAddressList(msg)
+	for field, addrs := range addr_fields {
+		if len(addrs) > 0 {
+			mainHeader.Set(field, strings.Join(addrs, ","))
+		}
+	}
+
+	// References headers
+	mainHeader.Set("Message-ID", "<"+b.NewMessageId(em.Message.Message_id.Bytes())+">")
+
+	// Formatted multipart body
+	body, err := formatBody(msg, mainHeader)
+	if err != nil {
+		return err
+	}
+	em.Email.Raw = body
+	log.Debug("Raw encrypted mail", em.Email.Raw)
+
+	json_rep, _ := EmailToJsonRep(em.Email.Raw.String())
+	em.Email_json = &json_rep
+	return
+}
+
+// Format a participant into a mail.Address
+func formatAddress(participant Participant) *mail.Address {
+	addr := &mail.Address{Name: participant.Label, Address: participant.Address}
+	return addr
+}
+
+// Create mail.Address list indexed by participant type
+func formatAddressList(msg *Message) map[string][]string {
+	// Create address headers
+	addr_fields := newAddressesFields()
+	for _, participant := range msg.Participants {
+		address := formatAddress(participant)
+		switch participant.Type {
+		case ParticipantFrom:
+			addr_fields["From"] = append(addr_fields["From"], address.String())
```

Comment:
addresses formatting is not consistent with MarshalEmail formatting rule : miss embedding participant.Label. 
instead, why not make use of `addr_fields` previously computed by MarshalEmail as it's a map[string]string ?

Reply:
I use participant.Label when constructing the mail.Address structure, I think this formatting is more robust. And I can't use any structure of gomail.v2 package, so this is why I can't use same mechanism than for a non encrypted mail (use of mail.Message.FormatAddress)

Reply:
`addr_fields` is a map[string]string, not a gomail.v2 struct and it is already computed before the call to `MarshalEncryptedEmail`. You could call `MarshalEncryptedEmail` with one more params : 
`MarshalEncryptedEmail(msg *Message, em *EmailMessage, addr_fields map[string]string) error {` and add string from addr_fields map instead of `address.String()`